src/app/components/date-picker/date-picker.component.ts

```diff
@@ -0,0 +1,400 @@
+import { Component, Input, Output, EventEmitter } from '@angular/core'
+import { DatePipe } from '@angular/common'
+import { Observable } from 'rxjs/Observable'
+import { Subject } from 'rxjs/Subject'
+import { range } from 'lodash'
+
+import 'rxjs/add/operator/map'
+import 'rxjs/add/operator/publishReplay'
+import 'rxjs/add/operator/scan'
+import 'rxjs/add/operator/startWith'
+
+import { locales } from './locales'
+
+export enum Mode {
+  Day,
+  Month,
+  Year
+}
+
+export enum Action {
+  MoveLeft,
+  MoveRight,
+  ChangeMode,
+  Today,
+  Tomorrow,
+  Clear,
+  ClickCell
+}
+
+@Component({
+  selector: 'date-picker',
+  directives: [
+  ],
+  templateUrl: './date-picker.template.html',
+  styleUrls: [ './date-picker.style.css' ]
+})
+export class DatePicker {
+
+  @Input() startDate: Date = new Date()
+  @Input() language: string
+
+  @Output() selectDate: EventEmitter<Date> = new EventEmitter<Date>()
+  @Output() onClickToday: EventEmitter<Date> = new EventEmitter<Date>()
+  @Output() onClickTomorrow: EventEmitter<Date> = new EventEmitter<Date>()
+  @Output() onClickClear: EventEmitter<Date> = new EventEmitter<Date>()
+
+  private YEAR_RANGE = 12
+  private DEFAULT_LANG = 'en'
+  private locale
+
+  private action$: Subject<any>
+  private state$: Observable<any>
+  private title$: Observable<string>
+  private rows$: Observable<Array<any>>
+
+  ngOnInit() {
+    this.language = this.language in locales ? this.language : this.DEFAULT_LANG
+    this.locale = locales[this.language]
+
+    let defaultState = {
+      currentDate: this.startDate,
+      currentMode: Mode.Day,
+      step: this.getStep(Mode.Day)
+    }
+
+    this.action$ = new Subject<any>()
+
+    this.state$ = this.action$
+      .startWith(defaultState)
+      .scan((state, action) => {
+        let newState = this.reducer(state, action)
+        this.sideEffect(newState, action)
+        return newState
+      })
+
+    this.title$ = this.state$
+      .map((state) => this.getTitle(state))
```

Comment:
We could have used `.map(this.getTitle.bind(this))`, but doing this will lose type signature, resulting a typescript warning.